drivers: video: CONFIG_VIDEO_SW_GENERATOR -> zephyr,video-sw-generator

josuah opened the pull request:
Downstream:

- https://github.com/zephyrproject-rtos/zephyr/pull/76798
- https://github.com/zephyrproject-rtos/zephyr/pull/88780

Related:

- https://github.com/zephyrproject-rtos/zephyr/pull/76798#discussion_r2019759097
- https://github.com/zephyrproject-rtos/zephyr/pull/87869
- https://github.com/zephyrproject-rtos/zephyr/pull/81725
- https://github.com/zephyrproject-rtos/zephyr/pull/88182

As pointed out in the UVC PR, the typical way to configure hardware-like devices in Zephyr is via a [chosen node](https://docs.zephyrproject.org/latest/build/dts/api/api.html#zephyr-specific-chosen-nodes). The other proposed approach for default video devices is #88182.

Because this is a purely virtual hardware, a snippet is provided to enable it in any example via `--snippet video-sw-generator`. Document its use in the 3 capture samples `README.rst`. The [`zephyr,camera` doc](https://builds.zephyrproject.io/zephyr/pr/88780/docs/build/dts/api/api.html#zephyr-specific-chosen-nodes) was missing and is also added.

If this is indeed the way to go, I will convert these other PRs to use the devicetree instead of Kconfig:

- https://github.com/zephyrproject-rtos/zephyr/pull/87007
- https://github.com/zephyrproject-rtos/zephyr/pull/87008

josuah committed:
drivers: video: sw_generator: convert to use the devicetree

Make the video software generator a devicetree node, which allows enabling
several instances, and select it as chosen { zephyr,camera = &... }; node.
It can be enabled via a `video-sw-generator` snippet.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: convert to new LOG_MODULE_REGISTER() syntax

Convert the log module declarationto the new syntax introduced in
5e34681 proposing a more compact syntax.
This fixes the log level not being updated in the "capture" sample.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: tcpserversink: add _dev suffix to variable

Variables that holds pointers to 'const struct device' typically end-up
with '_dev'. The tcpserversink video sample did not have it. Add it.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
doc: build: dts: api: document the zephyr,camera chosen node

The zephyr,camera was used in several samples, but was not documented.
Add the missing documentation entry for it in the list.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
I picked `video-sw-generator` for the node label following https://github.com/zephyrproject-rtos/zephyr/issues/53506
Maybe `zephyr_video_sw_generator` is more appropriate...

loicpoulain reviewed (dismissed):
LGTM, but not tested.

ngphibang commented:
It's fine to me to turn video sw generator into devicetree so that https://github.com/zephyrproject-rtos/zephyr/pull/76798 can get merged because it's been there for a long time.

For a long-term solution, as discussed in https://github.com/zephyrproject-rtos/zephyr/pull/88182 and in https://github.com/zephyrproject-rtos/zephyr/pull/76798:

- It's weird that to be used as a camera, we have to convert all the SW emulated driver into devicetree, as naturally these devices are not HW and not need to be described via a devicetree.
- `zephyr,camera` chosen node only specifies a unique device as camera while we should support multiple cameras at the same time: multiple HW camera sensors, or HW cameras and SW emulated devices and an ability to switch between them.

But we can do that later as an extension to not postpone https://github.com/zephyrproject-rtos/zephyr/pull/76798.

josuah commented on dts/bindings/video/zephyr,video-sw-generator.yaml:
```diff
@@ -0,0 +1,12 @@
+
+include: base.yaml
+
+child-binding:
+  child-binding:
+    include: video-interfaces.yaml
```
Good point, no "sub-devices"... It is the "head" of the pipeline on its own.

ngphibang replied:
It seems you forgot to remove these ?

josuah replied:
Right thank you! I thought I needed to use it for the [`video_sw_pipeline`](https://github.com/zephyrproject-rtos/zephyr/pull/88839) but it ended-up not the case, I simply did not know how to best handle "m2m" devices. removed now!

josuah committed:
samples: drivers: video: fix board definitions

By having an "auto-fallback" mechanism, the VIDEO_SW_GENERATOR did shadow
the fact that "platform:mimxrt1064_evk:SHIELD=dvp_fpc24_mt9m114" was
missing the full specification, and therefore was not matched at all:
the shield was not selected, but the CI still worked.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- Fix CI YAML bug hidden by the auto-fallback behavior of `VIDEO_SW_GENERATOR`.

josuah commented:
> But we can do that later as a future extension (as currently, we are all based on zephyr,camera) to not postpone #76798.

#76798 (UVC) is not a dependency but thank you for caring.

Hopefully by switching to `zephyr,camera` first, it will be easier for reviewers of #88182 to understand the limitations of this approach, as they would immediately face it.

josuah commented:
Drafting until rebase...

JarmouniA reviewed (commented):
Non-blocking

josuah commented on samples/drivers/video/capture/README.rst:
```diff
@@ -72,7 +72,14 @@ commands:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/video/capture
+   :board: native_sim
```
Good point, a lot easier when not needing to install 32-bit libraries for SDL.

josuah committed:
samples: drivers: video: tcpserversink: sort header alphabetically

Since this sample's headers are now grouped it is possible to sort them
alphabetically.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented on samples/drivers/video/capture_to_lvgl/README.rst:
```diff
@@ -37,6 +37,16 @@ For :zephyr:board:`mini_stm32h743`, build this sample application with the follo
    :gen-args: -DCONFIG_BOOT_DELAY=2000
    :compact:
 
+For testing purpose without the need of any real video capture hardware,
```
Ah, my bad I forgot to apply this to *each* sample...

josuah commented on dts/bindings/video/zephyr,video-sw-generator.yaml:
```diff
@@ -0,0 +1,8 @@
+
+compatible: "zephyr,video-sw-generator"
+
+include: base.yaml
```
No, that is correct. I added that assuming that was required for any node, but if it can work without it, probably the way to go!

josuah commented:
Thanks for the review @avolmat-st! Though I just applied @JarmouniA diff saying `include: base.yaml` was not needed after all.

avolmat-st commented on samples/drivers/video/capture_to_lvgl/src/main.c:
```diff
@@ -15,12 +15,11 @@
 {
 	struct video_buffer *buffers[2], *vbuf;
 	const struct device *display_dev;
+	const struct device *video_dev;
```
video_dev already declared few lines below. Cf twister build failure

josuah replied:
Yes, this will need another round... Thank you for pointing at it.

To avoid running into so much issues, I am preparing a test rig. I am not yet to the point of running automated twister batches but at least I can re-run tests and samples...

My bad @JarmouniA I dismissed your review once again
